docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+
+      |Byte 0 (validity bitmap) | Bytes 1-63            |
+      |-------------------------|-----------------------|
+      | 00011101                | 0 (padding)           |
+
+    * Value Buffer:
+
+      |Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes 12-15 | Bytes 16-19 | Bytes 20-63 |
+      |------------|-------------|-------------|-------------|-------------|-------------|
+      | 1          | unspecified | 2           | 4           | 8           | unspecified |
+
+**Example Layout: Non-null int32 Array**
+
+``[1, 2, 3, 4, 8]`` has two possible layouts: ::
+
+    * Length: 5, Null count: 0
+    * Null bitmap buffer:
+
+      | Byte 0 (validity bitmap) | Bytes 1-63            |
+      |--------------------------|-----------------------|
+      | 00011111                 | 0 (padding)           |
+
+    * Value Buffer:
+
+      |Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | bytes 12-15 | bytes 16-19 | Bytes 20-63 |
+      |------------|-------------|-------------|-------------|-------------|-------------|
+      | 1          | 2           | 3           | 4           | 8           | unspecified |
+
+or with the bitmap elided: ::
+
+    * Length 5, Null count: 0
+    * Null bitmap buffer: Not required
+    * Value Buffer:
+
+      |Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | bytes 12-15 | bytes 16-19 | Bytes 20-63 |
+      |------------|-------------|-------------|-------------|-------------|-------------|
+      | 1          | 2           | 3           | 4           | 8           | unspecified |
+
+Variable-size
+-------------
+
+Each value in this layout type consists of 0 or more bytes. While
+primitive arrays have a single values buffer, variable-size binary
+have an **offsets** buffer and **data** buffer
+
+The offsets buffer contains `length + 1` signed integers (either
+32-bit or 64-bit), which encode the start position of each element in
+the data buffer. The length of the value in each slot is computed
+using the first difference with the next element in the offsets
+array. For example, the position and length of slot j is computed as:
+
+::
+
+    slot_position = offsets[j]
+    slot_length = offsets[j + 1] - offsets[j]  // (for 0 <= j < length)
+
+Generally the first value in the offsets array is 0, and the last
+element is the length of the values array.
+
+Variable-size List
+------------------
+
+List is a nested type which is semantically similar to variable-size
+binary. It is defined by two buffers: a validity bitmap and an offsets
+buffer. The offsets are the same as in the variable-size binary case,
+and both 32-bit and 64-bit signed integer offsets are
+supported. Rather than referencing an additional data buffer, instead
+these offsets reference a child array having any type.
+
+A list type is specified like ``List<T>``, where ``T`` is any type
+(primitive or nested).
+
+**Example Layout: ``List<Char>`` Array**
+
+We illustrate ``List<Int8>``.
+
+For an array of length 4 with respective values: ::
```

Comment:
Should it be "An array of length 4 with respective values (...) will have the following representation"?